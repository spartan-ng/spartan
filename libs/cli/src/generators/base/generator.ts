import {
	type GeneratorCallback,
	type Tree,
	addDependenciesToPackageJson,
	generateFiles,
	joinPathFragments,
	runTasksInSerial,
} from '@nx/devkit';
import { addTsConfigPath } from '@nx/js';
import * as path from 'node:path';
import { readTsConfigPathsFromTree } from '../../utils/tsconfig';
import { getInstalledPackageVersion } from '../../utils/version-utils';
import { buildDependencyArray, buildDevDependencyArray } from './lib/build-dependency-array';
import { deleteFiles } from './lib/deleteFiles';
import { getTargetLibraryDirectory } from './lib/get-target-library-directory';
import { initializeAngularLibrary } from './lib/initialize-angular-library';
import type { HlmBaseGeneratorSchema } from './schema';
import { FALLBACK_ANGULAR_VERSION } from './versions';

export async function hlmBaseGenerator(tree: Tree, options: HlmBaseGeneratorSchema) {
	const tasks: GeneratorCallback[] = [];
	const targetLibDir = getTargetLibraryDirectory(options, tree);

	const existingPathsByAlias = readTsConfigPathsFromTree(tree);
	const tsConfigAliasToUse = `@spartan-ng/${options.publicName}`;

	if (Object.keys(existingPathsByAlias).includes(tsConfigAliasToUse)) {
		console.log(`Skipping ${tsConfigAliasToUse}. It's already installed!`);
		return runTasksInSerial(...tasks);
	}

	if (options.angularCli) {
		addTsConfigPath(tree, tsConfigAliasToUse, [`.${path.sep}${joinPathFragments(targetLibDir, 'src', 'index.ts')}`]);
	} else {
		tasks.push(await initializeAngularLibrary(tree, options));
	}

	// we need to delete the src/lib/{publicName} directory because it is generated by nx and cannot be disabled
	deleteFiles(tree, joinPathFragments(options.directory, options.publicName, 'src', 'lib', options.publicName));

	// does add the files to the repository
	generateFiles(
		tree,
		path.join(__dirname, '..', 'ui', 'libs', options.internalName, 'files'),
		path.join(targetLibDir, 'src'),
		options,
	);

	const angularVersion = getInstalledPackageVersion(tree, '@angular/core', FALLBACK_ANGULAR_VERSION, true);
	const existingCdkVersion = getInstalledPackageVersion(tree, '@angular/cdk', FALLBACK_ANGULAR_VERSION, true);
	const dependencies = buildDependencyArray(options, angularVersion, existingCdkVersion);
	const devDependencies = buildDevDependencyArray();

	tasks.push(addDependenciesToPackageJson(tree, dependencies, devDependencies));
	return runTasksInSerial(...tasks);
}

export default hlmBaseGenerator;
