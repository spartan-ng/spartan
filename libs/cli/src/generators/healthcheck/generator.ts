import { formatFiles, logger, Tree } from '@nx/devkit';
import { Healthcheck, HealthcheckReport, HealthcheckStatus, isHealthcheckFixable } from './healthchecks';
import { brainImportsHealthcheck } from './healthchecks/brain-imports';
import { brainRadioHealthcheck } from './healthchecks/brn-radio';
import { coreImportsHealthcheck } from './healthchecks/core-imports';
import { helmIconHealthcheck } from './healthchecks/hlm-icon';
import { scrollAreaHealthcheck } from './healthchecks/hlm-scroll-area';
import { versionHealthcheck } from './healthchecks/version';
import { HealthcheckGeneratorSchema } from './schema';
import { promptUser } from './utils/prompt';
import { printReport } from './utils/reporter';
import { runHealthcheck } from './utils/runner';

export async function healthcheckGenerator(tree: Tree, options: HealthcheckGeneratorSchema) {
	logger.info('Running healthchecks...');

	const healthchecks: Healthcheck[] = [
		versionHealthcheck,
		brainImportsHealthcheck,
		coreImportsHealthcheck,
		helmIconHealthcheck,
		scrollAreaHealthcheck,
		brainRadioHealthcheck,
	];

	// store all the failed healthchecks that can be fixed
	const failedReports: HealthcheckReport[] = [];

	for (const healthcheck of healthchecks) {
		const report = await runHealthcheck(tree, healthcheck);
		printReport(report);

		if (report.status === HealthcheckStatus.Failure) {
			failedReports.push(report);
		}
	}

	// if there are some failed healthchecks that can be fixed, ask the user if they want to fix them
	for (const report of failedReports) {
		if (report.fixable && isHealthcheckFixable(report.healthcheck)) {
			const fix = options.autoFix || (await promptUser(report.healthcheck.prompt));

			if (fix) {
				await report.healthcheck.fix(tree);
			}
		}
	}

	if (!options.skipFormat) {
		await formatFiles(tree);
	}
}

export default healthcheckGenerator;
