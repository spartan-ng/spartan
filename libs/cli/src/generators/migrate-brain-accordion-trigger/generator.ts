import { formatFiles, type Tree } from '@nx/devkit';
import { NodeType } from 'node-html-parser';
import { Project } from 'ts-morph';
import { visitFiles } from '../../utils/visit-files';
import type { MigrateAccordionTriggerGeneratorSchema } from './schema';
import {
	findAccordionButtons,
	type HTMLElement,
	parseHtmlForAccordionMigration,
	wrapButtonInHeading,
} from './utils/html-utils';
import {
	analyzeButton,
	type ButtonAnalysis,
	extractInlineTemplate,
	hasAccordionTriggers,
	shouldProcessFile,
} from './utils/shared-utils';

export default async function migrateAccordionTriggerGenerator(
	tree: Tree,
	{ skipFormat }: MigrateAccordionTriggerGeneratorSchema,
) {
	const issues: string[] = [];
	const project = new Project({ useInMemoryFileSystem: true });

	wrapAccordionTriggers(tree, project, issues);

	if (issues.length > 0) {
		reportManualFixesNeeded(issues);
	}

	if (!skipFormat) {
		await formatFiles(tree);
	}
}

function wrapAccordionTriggers(tree: Tree, project: Project, issues: string[]) {
	visitFiles(tree, '.', (path) => {
		console.log('Processing ' + path + '...');
		if (!shouldProcessFile(path)) return;

		const content = tree.read(path, 'utf-8');
		if (!content || !hasAccordionTriggers(content)) return;

		const result = path.endsWith('.html')
			? fixTemplate(content, path, issues)
			: fixComponentFile(content, path, project, issues);

		if (result?.modified) {
			tree.write(path, result.content);
			console.log(`Fixed ${result.fixedCount} accordion trigger(s) in ${path}`);
		}
	});
}

interface FixResult {
	content: string;
	modified: boolean;
	fixedCount: number;
}

function fixTemplate(
	content: string,
	filePath: string,
	issues: string[],
	options: { noContentsClass: boolean } = { noContentsClass: false },
): FixResult {
	const root = parseHtmlForAccordionMigration(content);
	const buttons = findAccordionButtons(root);
	let fixedCount = 0;

	for (const button of buttons) {
		const analysis = analyzeButton(button);

		switch (analysis.issue) {
			case 'NotInHeading':
				wrapButtonInHeading(button, !options.noContentsClass);
				fixedCount++;
				break;

			case 'HasSiblings':
				logSiblingIssue(analysis, filePath, issues);
				break;

			case 'NoIssue':
				break;
		}
	}

	if (fixedCount > 0) {
		return {
			content: formatHtmlOutput(root),
			modified: true,
			fixedCount,
		};
	}

	return { content, modified: false, fixedCount: 0 };
}

function fixComponentFile(content: string, filePath: string, project: Project, issues: string[]): FixResult {
	const sourceFile = project.createSourceFile(filePath, content, { overwrite: true });

	const templateInfo = extractInlineTemplate(sourceFile);
	if (!templateInfo) {
		return { content, modified: false, fixedCount: 0 };
	}

	// Fix the template content
	const result = fixTemplate(templateInfo.text, filePath, issues);

	if (result.modified && templateInfo.propertyAssignment.getInitializer()) {
		// Replace template in source file
		const quoteStyle = templateInfo.originalQuoteStyle || '`';
		let newTemplateValue: string;

		if (quoteStyle === '`') {
			newTemplateValue = `\`${result.content}\``;
		} else if (quoteStyle === '"') {
			newTemplateValue = `"${result.content.replace(/"/g, '\\"')}"`;
		} else {
			newTemplateValue = `'${result.content.replace(/'/g, "\\'")}'`;
		}

		templateInfo.propertyAssignment.getInitializer().replaceWithText(newTemplateValue);

		return {
			content: sourceFile.getFullText(),
			modified: true,
			fixedCount: result.fixedCount,
		};
	}

	return { content, modified: false, fixedCount: 0 };
}

function logSiblingIssue(analysis: ButtonAnalysis, filePath: string, issues: string[]) {
	const parent = analysis.parent;
	const siblingContent = getSiblingPreview(analysis.button, parent);

	issues.push(
		`! ${filePath}\n` +
			` - Button "${analysis.buttonText}" in <${parent.tagName.toLowerCase()}> ` +
			`has siblings that need to be moved: ${siblingContent}...\n`,
	);
}

function getSiblingPreview(button: HTMLElement, parent: HTMLElement): string {
	return parent.childNodes
		.filter((child) => child !== button)
		.map((child) => {
			if (child.nodeType === NodeType.TEXT_NODE) {
				return child.text?.trim();
			}
			return child.toString();
		})
		.filter(Boolean)
		.join(', ')
		.substring(0, 50);
}

function formatHtmlOutput(root: HTMLElement): string {
	const output = root.toString();
	// Convert empty tags to self-closing
	return output.replace(/<([^>\s]+)([^>]*)><\/\1>/g, '<$1$2/>');
}

function reportManualFixesNeeded(issues: string[]) {
	console.log(
		`We also detected ${issues.length} accordion trigger buttons ` +
			`that have siblings that need to be manually moved:`,
	);
	issues.forEach((issue) => console.log(issue));
	console.log('To learn more about this pattern see: ' + 'https://www.w3.org/WAI/ARIA/apg/patterns/accordion/');
}
