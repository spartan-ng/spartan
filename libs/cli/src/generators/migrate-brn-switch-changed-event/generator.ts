import { formatFiles, type Tree } from '@nx/devkit';
import { visitFiles } from '../../utils/visit-files';
import type { MigrateBrnSwitchChangedEvent } from './schema';

export async function migrateBrnSwitchChangedEvent(tree: Tree, { skipFormat }: MigrateBrnSwitchChangedEvent) {
	replaceOpenChangeEvent(tree);

	if (!skipFormat) {
		await formatFiles(tree);
	}
}

function replaceOpenChangeEvent(tree: Tree) {
	// if the element is `'<hlm-select' and it has an `(openedChange)` event, we need to replace it with `(openChange)`
	visitFiles(tree, '.', (path) => {
		// if this is not an html file or typescript file (inline templates) then skip
		if (!path.endsWith('.html') && !path.endsWith('.ts')) {
			return;
		}

		let content = tree.read(path, 'utf-8');

		if (!content) {
			return;
		}

		// find all the brn-select or hlm-select elements that have an `(openedChange)` event
		content = content.replace(/<(brn-switch)[^>]*\(\s*changed\s*\)=/g, (match) =>
			match.replace(/\(\s*changed\s*\)/, '(checkedChange)'),
		);

		tree.write(path, content);
	});
}

export default migrateBrnSwitchChangedEvent;
