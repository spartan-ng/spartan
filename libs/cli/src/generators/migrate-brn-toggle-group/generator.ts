import { formatFiles, type Tree } from '@nx/devkit';
import { visitFiles } from '../../utils/visit-files';
import type { MigrateBrnToggleGroupGeneratorSchema } from './schema';

export async function migrateBrnToggleGroupGenerator(tree: Tree, { skipFormat }: MigrateBrnToggleGroupGeneratorSchema) {
	updateImports(tree);
	replaceSelector(tree);

	if (!skipFormat) {
		await formatFiles(tree);
	}
}

/**
 * Update imports remove BrnToggleGroup import
 */
function updateImports(tree: Tree) {
	visitFiles(tree, '/', (path) => {
		// skip HlmToggleGroup itself
		if (path.endsWith('hlm-toggle-group.ts') || path.endsWith('hlm-toggle-group-item.ts')) {
			return;
		}

		const content = tree.read(path, 'utf-8');

		if (!content) {
			return;
		}

		if (content.includes("'@spartan-ng/brain/toggle-group';") || content.includes("'@spartan-ng/helm/toggle-group';")) {
			const updatedContent = content
				// Handle `import { BrnToggleGroup, BrnToggleGroupItem } from '@spartan-ng/brain/toggle-group';`
				.replace("import { BrnToggleGroup, BrnToggleGroupItem } from '@spartan-ng/brain/toggle-group';", '')
				// Handle `import { BrnToggleGroup } from '@spartan-ng/brain/toggle-group';`
				.replace("import { BrnToggleGroup } from '@spartan-ng/brain/toggle-group';", '')
				// Handle `import { BrnToggleGroupItem } from '@spartan-ng/brain/toggle-group';`
				.replace("import { BrnToggleGroupItem } from '@spartan-ng/brain/toggle-group';", '')
				// Remove `BrnToggleGroupItem` with optional comma and whitespace
				.replace(/BrnToggleGroupItem,?\s?/, '')
				// Remove `BrnToggleGroup` with optional comma and whitespace
				.replace(/BrnToggleGroup,?\s?/, '');

			tree.write(path, updatedContent);
		}
	});
}

function replaceSelector(tree: Tree) {
	// if the element is `<brn-toggle-group hlm` then we need to replace it with `<hlm-toggle-group`
	visitFiles(tree, '.', (path) => {
		// if this is not an html file or typescript file (inline templates) then skip
		if (!path.endsWith('.html') && !path.endsWith('.ts')) {
			return;
		}

		let content = tree.read(path, 'utf-8');

		if (!content) {
			return;
		}

		if (content.includes('brn-toggle-group') || content.includes('@spartan-ng/helm/toggle-group')) {
			// <brn-toggle-group hlm but between
			content = replaceBrnToggleGroup(content);
			// replace closing tag
			content = content.replace(/<\/brn-toggle-group>/g, '</hlm-toggle-group>');

			tree.write(path, content);
		}
	});
}

function replaceBrnToggleGroup(input) {
	// Split input to handle multiple tags separately
	return input
		.split(/(?=<)/)
		.map((tag) => {
			// Skip if not a brn-toggle-group tag
			if (!tag.startsWith('<brn-toggle-group')) {
				return tag;
			}

			// Remove line breaks, tabs, and other whitespace within the tag
			// Replace with a single space
			tag = tag.replace(/\s+/g, ' ');

			// Check if standalone hlm attribute exists
			const hasHlm = / hlm(?=[\s>])/.test(tag);

			if (hasHlm) {
				// Remove the hlm attribute and convert to hlmToggle
				return tag
					.replace(/<brn-toggle-group/, '<hlm-toggle-group')
					.replace(/ hlm(?=[\s>])/, '')
					.replace(/ multiple="false"/, ' type="single"')
					.replace(/ multiple="true"/, ' type="multiple"')
					.replace(/\s+>/g, '>')
					.replace(/\s+/g, ' ');
			}

			return tag;
		})
		.join('');
}

export default migrateBrnToggleGroupGenerator;
