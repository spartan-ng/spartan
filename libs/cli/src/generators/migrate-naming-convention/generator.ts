import { formatFiles, Tree } from '@nx/devkit';
import { tsquery } from '@phenomnomnominal/tsquery';
import * as ts from 'typescript';
import { visitFiles } from '../../utils/visit-files';
import { MigrateNamingConventionsGeneratorSchema } from './schema';

export async function migrateNamingConventionGenerator(
	tree: Tree,
	{ skipFormat }: MigrateNamingConventionsGeneratorSchema,
) {
	updateBrainIdentifiers(tree);
	updateHelmIdentifiers(tree);

	if (!skipFormat) {
		await formatFiles(tree);
	}
}

const renamedIdentifiers: Record<string, string> = {
	BrnTooltipContentDirective: 'BrnTooltipContentTemplate',
	BrnSelectValueDirective: 'BrnSelectValueTemplate',
	HlmSelectValueDirective: 'HlmSelectValueTemplate',
	HlmMenuItemRadioComponent: 'HlmMenuItemRadioIndicator',
};

/**
 * Detect any imports from `@spartan-ng/brain/*` that need to be updated.
 * Get the identifiers that need to be replaced, and then replace them.
 */
function updateBrainIdentifiers(tree: Tree) {
	visitFiles(tree, '/', (path) => {
		if (!path.endsWith('.ts')) {
			return;
		}

		let content = tree.read(path).toString('utf-8');

		const importDeclarations = tsquery.query<ts.ImportDeclaration>(content, 'ImportDeclaration');

		if (importDeclarations.length === 0) {
			return;
		}

		const identifiersToReplace: string[] = [];

		for (const importDeclaration of importDeclarations) {
			const moduleSpecifier = importDeclaration.moduleSpecifier.getText();

			if (moduleSpecifier.includes('@spartan-ng/brain/')) {
				const identifiers = tsquery.query(importDeclaration, 'Identifier');

				// we only care about identifiers that end in 'Directive', 'Component'
				for (const identifier of identifiers) {
					const identifierText = identifier.getText();

					if (identifierText.endsWith('Directive') || identifierText.endsWith('Component')) {
						identifiersToReplace.push(identifierText);
					}
				}
			}
		}

		// find all the identifiers in the file - we reverse the order to replace the exact identifiers based on position
		const identifiers = tsquery.query<ts.Identifier>(content, 'Identifier').reverse();

		for (const identifier of identifiers) {
			if (!identifiersToReplace.includes(identifier.getText())) {
				continue;
			}

			// determine the new identifier based on the old one, if it exists in the mapping use it, otherwise just remove
			// the trailing 'Directive' or 'Component'
			const oldIdentifier = identifier.getText();
			const newIdentifier = renamedIdentifiers[oldIdentifier] || oldIdentifier.replace(/(Directive|Component)$/, '');

			if (newIdentifier !== oldIdentifier) {
				const start = identifier.getStart();
				const end = identifier.getEnd();

				content = content.slice(0, start) + newIdentifier + content.slice(end);
			}
		}

		tree.write(path, content);
	});
}

/**
 * Helm identifiers are not necessarily just imports as the class declarations
 * are defined in the users codebase so we need to update all identifiers that
 * start with `Hlm` and end with `Directive` or `Component`.
 * We also must check for known identifiers that have been renamed.
 */
function updateHelmIdentifiers(tree: Tree) {
	visitFiles(tree, '/', (path) => {
		if (!path.endsWith('.ts')) {
			return;
		}

		let content = tree.read(path).toString('utf-8');

		const identifiers = tsquery.query<ts.Identifier>(content, 'Identifier').reverse();
		const identifiersToReplace: string[] = [];
		for (const identifier of identifiers) {
			const identifierText = identifier.getText();

			if (
				identifierText.startsWith('Hlm') &&
				(identifierText.endsWith('Directive') || identifierText.endsWith('Component'))
			) {
				identifiersToReplace.push(identifierText);
			}
		}

		// replace all identifiers in the content
		for (const oldIdentifier of identifiersToReplace) {
			const newIdentifier = renamedIdentifiers[oldIdentifier] || oldIdentifier.replace(/(Directive|Component)$/, '');

			content = content.replace(new RegExp(`\\b${oldIdentifier}\\b`, 'g'), newIdentifier);
		}

		tree.write(path, content);
	});
}

export default migrateNamingConventionGenerator;
