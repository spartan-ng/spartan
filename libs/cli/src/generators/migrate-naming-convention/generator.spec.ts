import { applicationGenerator, E2eTestRunner, UnitTestRunner } from '@nx/angular/generators';
import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { migrateNamingConventionGenerator } from './generator';

// patch some imports to avoid running the actual code
jest.mock('enquirer');
jest.mock('@nx/devkit', () => {
	const original = jest.requireActual('@nx/devkit');
	return {
		...original,
		ensurePackage: (pkg: string) => jest.requireActual(pkg),
		createProjectGraphAsync: jest.fn().mockResolvedValue({
			nodes: {},
			dependencies: {},
		}),
	};
});

describe('migrate-naming-conventions generator', () => {
	let tree: Tree;

	beforeEach(async () => {
		tree = createTreeWithEmptyWorkspace();

		await applicationGenerator(tree, {
			name: 'app',
			directory: 'app',
			skipFormat: true,
			e2eTestRunner: E2eTestRunner.None,
			unitTestRunner: UnitTestRunner.None,
			skipPackageJson: true,
			skipTests: true,
		});
	});

	it('should not replace helm identifiers', async () => {
		tree.write(
			'app/src/app/app.component.ts',
			`import { Component } from '@angular/core';
import { NgIcon } from '@ng-icons/core';
import {
	HlmAccordionContentComponent,
	HlmAccordionDirective,
	HlmAccordionIconDirective,
	HlmAccordionItemDirective,
	HlmAccordionTriggerDirective,
} from '@spartan-ng/helm/accordion';
import { HlmIconDirective } from '@spartan-ng/helm/icon';

@Component({
	selector: 'spartan-accordion-preview',
	imports: [
		HlmAccordionDirective,
		HlmAccordionItemDirective,
		HlmAccordionTriggerDirective,
		HlmAccordionIconDirective,
		HlmAccordionContentComponent,
		NgIcon,
		HlmIconDirective,
	],
	template: \`
		<div hlmAccordion>
			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Product Information
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					Our flagship product combines cutting-edge technology with sleek design. Built with premium materials, it
					offers unparalleled performance and reliability.

					<br />
					<br />

					Key features include advanced processing capabilities, and an intuitive user interface designed for both
					beginners and experts.
				</hlm-accordion-content>
			</div>

			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Shipping Details
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					We offer worldwide shipping through trusted courier partners. Standard delivery takes 3-5 business days, while
					express shipping ensures delivery within 1-2 business days.

					<br />
					<br />
					All orders are carefully packaged and fully insured. Track your shipment in real-time through our dedicated
					tracking portal.
				</hlm-accordion-content>
			</div>

			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Return Policy
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					We stand behind our products with a comprehensive 30-day return policy. If you're not completely satisfied,
					simply return the item in its original condition.
					<br />
					<br />
					Our hassle-free return process includes free return shipping and full refunds processed within 48 hours of
					receiving the returned item.
				</hlm-accordion-content>
			</div>
		</div>
	\`,
})
export class AccordionPreviewComponent {}`,
		);

		await migrateNamingConventionGenerator(tree, { skipFormat: true });

		const content = tree.read('app/src/app/app.component.ts', 'utf-8');
		expect(content).toMatchSnapshot();
	});

	it('should replace brain identifiers', async () => {
		tree.write(
			'app/src/app/app.component.ts',
			`
import { Component, signal } from '@angular/core';
import { NgIcon, provideIcons } from '@ng-icons/core';
import { lucideCheck, lucideChevronsUpDown, lucideSearch } from '@ng-icons/lucide';
import { BrnCommandImports } from '@spartan-ng/brain/command';
import { BrnPopoverComponent, BrnPopoverContentDirective, BrnPopoverTriggerDirective } from '@spartan-ng/brain/popover';
import { HlmButtonDirective } from '@spartan-ng/helm/button';
import { HlmCommandImports } from '@spartan-ng/helm/command';
import { HlmIconDirective } from '@spartan-ng/helm/icon';
import { HlmPopoverContentDirective } from '@spartan-ng/helm/popover';

type Framework = { label: string; value: string };

@Component({
	selector: 'spartan-combobox-preview',
	imports: [
		BrnCommandImports,
		HlmCommandImports,
		NgIcon,
		HlmIconDirective,
		HlmButtonDirective,
		BrnPopoverComponent,
		BrnPopoverTriggerDirective,
		HlmPopoverContentDirective,
		BrnPopoverContentDirective,
	],
	providers: [provideIcons({ lucideChevronsUpDown, lucideSearch, lucideCheck })],
	template: \`
		<brn-popover [state]="state()" (stateChanged)="stateChanged($event)" sideOffset="5">
			<button
				class="w-[200px] justify-between"
				id="edit-profile"
				variant="outline"
				brnPopoverTrigger
				(click)="state.set('open')"
				hlmBtn
			>
				{{ currentFramework() ? currentFramework()?.label : 'Select framework...' }}
				<ng-icon hlm size="sm" name="lucideChevronsUpDown" class="opacity-50" />
			</button>
			<hlm-command *brnPopoverContent="let ctx" hlmPopoverContent class="w-[200px] p-0">
				<hlm-command-search>
					<ng-icon hlm name="lucideSearch" />
					<input placeholder="Search framework..." hlm-command-search-input />
				</hlm-command-search>
				<div *brnCommandEmpty hlmCommandEmpty>No results found.</div>
				<hlm-command-list>
					<hlm-command-group>
						@for (framework of frameworks; track framework) {
							<button hlm-command-item [value]="framework.value" (selected)="commandSelected(framework)">
								<span>{{ framework.label }}</span>
								<ng-icon
									hlm
									class="ml-auto"
									[class.opacity-0]="currentFramework()?.value !== framework.value"
									name="lucideCheck"
									hlmCommandIcon
								/>
							</button>
						}
					</hlm-command-group>
				</hlm-command-list>
			</hlm-command>
		</brn-popover>
	\`,
})
export class ComboboxPreviewComponent {
	public frameworks = [
		{
			label: 'AnalogJs',
			value: 'analogjs',
		},
		{
			label: 'Angular',
			value: 'angular',
		},
		{
			label: 'Vue',
			value: 'vue',
		},
		{
			label: 'Nuxt',
			value: 'nuxt',
		},
		{
			label: 'React',
			value: 'react',
		},
		{
			label: 'NextJs',
			value: 'nextjs',
		},
	];
	public currentFramework = signal<Framework | undefined>(undefined);
	public state = signal<'closed' | 'open'>('closed');

	stateChanged(state: 'open' | 'closed') {
		this.state.set(state);
	}

	commandSelected(framework: Framework) {
		this.state.set('closed');
		if (this.currentFramework()?.value === framework.value) {
			this.currentFramework.set(undefined);
		} else {
			this.currentFramework.set(framework);
		}
	}
}
`,
		);

		await migrateNamingConventionGenerator(tree, { skipFormat: true });
		const content = tree.read('app/src/app/app.component.ts', 'utf-8');
		expect(content).toMatchSnapshot();
	});

	it('should rename the identifiers that changed name', async () => {
		tree.write(
			'app/src/app/app.component.ts',
			`import { Component } from '@angular/core';
import { BrnTooltipContentDirective } from '@spartan-ng/brain/tooltip';
import {HlmMenuItemRadioComponent} from '@spartan-ng/helm/menu';
import { BrnSelectValueDirective } from '@spartan-ng/brain/select';

@Component({
	selector: 'spartan-tooltip-preview',
	imports: [BrnTooltipContentDirective, HlmMenuItemRadioComponent, BrnSelectValueDirective],
	template: \`
		<brn-tooltip>
			<button brnTooltipTrigger>Hover me</button>
			<brn-tooltip-content>
				<p>This is a tooltip!</p>
			</brn-tooltip-content>
		</brn-tooltip>
	\`,
})
export class TooltipPreviewComponent {}`,
		);

		await migrateNamingConventionGenerator(tree, { skipFormat: true });

		const content = tree.read('app/src/app/app.component.ts', 'utf-8');
		expect(content).toMatchSnapshot();
	});
});
