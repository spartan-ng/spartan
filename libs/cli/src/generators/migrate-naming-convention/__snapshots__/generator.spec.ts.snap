// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrate-naming-conventions generator should not replace helm identifiers 1`] = `
"import { Component } from '@angular/core';
import { NgIcon } from '@ng-icons/core';
import {
	HlmAccordionContent,
	HlmAccordion,
	HlmAccordionIcon,
	HlmAccordionItem,
	HlmAccordionTrigger,
} from '@spartan-ng/helm/accordion';
import { HlmIcon } from '@spartan-ng/helm/icon';

@Component({
	selector: 'spartan-accordion-preview',
	imports: [
		HlmAccordion,
		HlmAccordionItem,
		HlmAccordionTrigger,
		HlmAccordionIcon,
		HlmAccordionContent,
		NgIcon,
		HlmIcon,
	],
	template: \`
		<div hlmAccordion>
			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Product Information
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					Our flagship product combines cutting-edge technology with sleek design. Built with premium materials, it
					offers unparalleled performance and reliability.

					<br />
					<br />

					Key features include advanced processing capabilities, and an intuitive user interface designed for both
					beginners and experts.
				</hlm-accordion-content>
			</div>

			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Shipping Details
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					We offer worldwide shipping through trusted courier partners. Standard delivery takes 3-5 business days, while
					express shipping ensures delivery within 1-2 business days.

					<br />
					<br />
					All orders are carefully packaged and fully insured. Track your shipment in real-time through our dedicated
					tracking portal.
				</hlm-accordion-content>
			</div>

			<div hlmAccordionItem>
				<button hlmAccordionTrigger>
					Return Policy
					<ng-icon name="lucideChevronDown" hlm hlmAccIcon />
				</button>
				<hlm-accordion-content>
					We stand behind our products with a comprehensive 30-day return policy. If you're not completely satisfied,
					simply return the item in its original condition.
					<br />
					<br />
					Our hassle-free return process includes free return shipping and full refunds processed within 48 hours of
					receiving the returned item.
				</hlm-accordion-content>
			</div>
		</div>
	\`,
})
export class AccordionPreviewComponent {}"
`;

exports[`migrate-naming-conventions generator should rename the identifiers that changed name 1`] = `
"import { Component } from '@angular/core';
import { BrnTooltipContentTemplate } from '@spartan-ng/brain/tooltip';
import {HlmMenuItemRadioIndicator} from '@spartan-ng/helm/menu';
import { BrnSelectValueTemplate } from '@spartan-ng/brain/select';

@Component({
	selector: 'spartan-tooltip-preview',
	imports: [BrnTooltipContentTemplate, HlmMenuItemRadioIndicator, BrnSelectValueTemplate],
	template: \`
		<brn-tooltip>
			<button brnTooltipTrigger>Hover me</button>
			<brn-tooltip-content>
				<p>This is a tooltip!</p>
			</brn-tooltip-content>
		</brn-tooltip>
	\`,
})
export class TooltipPreviewComponent {}"
`;

exports[`migrate-naming-conventions generator should replace brain identifiers 1`] = `
"
import { Component, signal } from '@angular/core';
import { NgIcon, provideIcons } from '@ng-icons/core';
import { lucideCheck, lucideChevronsUpDown, lucideSearch } from '@ng-icons/lucide';
import { BrnCommandImports } from '@spartan-ng/brain/command';
import { BrnPopover, BrnPopoverContent, BrnPopoverTrigger } from '@spartan-ng/brain/popover';
import { HlmButton } from '@spartan-ng/helm/button';
import { HlmCommandImports } from '@spartan-ng/helm/command';
import { HlmIcon } from '@spartan-ng/helm/icon';
import { HlmPopoverContent } from '@spartan-ng/helm/popover';

type Framework = { label: string; value: string };

@Component({
	selector: 'spartan-combobox-preview',
	imports: [
		BrnCommandImports,
		HlmCommandImports,
		NgIcon,
		HlmIcon,
		HlmButton,
		BrnPopover,
		BrnPopoverTrigger,
		HlmPopoverContent,
		BrnPopoverContent,
	],
	providers: [provideIcons({ lucideChevronsUpDown, lucideSearch, lucideCheck })],
	template: \`
		<brn-popover [state]="state()" (stateChanged)="stateChanged($event)" sideOffset="5">
			<button
				class="w-[200px] justify-between"
				id="edit-profile"
				variant="outline"
				brnPopoverTrigger
				(click)="state.set('open')"
				hlmBtn
			>
				{{ currentFramework() ? currentFramework()?.label : 'Select framework...' }}
				<ng-icon hlm size="sm" name="lucideChevronsUpDown" class="opacity-50" />
			</button>
			<hlm-command *brnPopoverContent="let ctx" hlmPopoverContent class="w-[200px] p-0">
				<hlm-command-search>
					<ng-icon hlm name="lucideSearch" />
					<input placeholder="Search framework..." hlm-command-search-input />
				</hlm-command-search>
				<div *brnCommandEmpty hlmCommandEmpty>No results found.</div>
				<hlm-command-list>
					<hlm-command-group>
						@for (framework of frameworks; track framework) {
							<button hlm-command-item [value]="framework.value" (selected)="commandSelected(framework)">
								<span>{{ framework.label }}</span>
								<ng-icon
									hlm
									class="ml-auto"
									[class.opacity-0]="currentFramework()?.value !== framework.value"
									name="lucideCheck"
									hlmCommandIcon
								/>
							</button>
						}
					</hlm-command-group>
				</hlm-command-list>
			</hlm-command>
		</brn-popover>
	\`,
})
export class ComboboxPreviewComponent {
	public frameworks = [
		{
			label: 'AnalogJs',
			value: 'analogjs',
		},
		{
			label: 'Angular',
			value: 'angular',
		},
		{
			label: 'Vue',
			value: 'vue',
		},
		{
			label: 'Nuxt',
			value: 'nuxt',
		},
		{
			label: 'React',
			value: 'react',
		},
		{
			label: 'NextJs',
			value: 'nextjs',
		},
	];
	public currentFramework = signal<Framework | undefined>(undefined);
	public state = signal<'closed' | 'open'>('closed');

	stateChanged(state: 'open' | 'closed') {
		this.state.set(state);
	}

	commandSelected(framework: Framework) {
		this.state.set('closed');
		if (this.currentFramework()?.value === framework.value) {
			this.currentFramework.set(undefined);
		} else {
			this.currentFramework.set(framework);
		}
	}
}
"
`;
